src/modules/safe-shield/contract-analysis/contract-analysis.service.ts

```diff
@@ -92,45 +99,52 @@ export class ContractAnalysisService {
     chainId: string;
     safeAddress: Address;
     contract: Address;
+    isDelegateCall: boolean;
   }): Promise<Record<ContractStatusGroup, Array<ContractAnalysisResult>>> {
-    const [contractVerificationResult, contractInteractionResult] =
+    const { chainId, safeAddress, contract, isDelegateCall } = args;
+
+    const [[verificationResult, delegateCallResult], interactionResult] =
       await Promise.all([
-        this.verifyContract({
-          chainId: args.chainId,
-          contract: args.contract,
-        }),
-        this.analyzeInteractions(args),
+        this.verifyContract({ chainId, contract, isDelegateCall }),
+        this.analyzeInteractions({ chainId, safeAddress, contract }),
       ]);
 
     return {
-      CONTRACT_VERIFICATION: [contractVerificationResult],
-      CONTRACT_INTERACTION: [contractInteractionResult],
-      DELEGATECALL: [],
+      CONTRACT_VERIFICATION: [verificationResult],
+      CONTRACT_INTERACTION: [interactionResult],
+      DELEGATECALL: delegateCallResult ? [delegateCallResult] : [],
     };
   }
 
   /**
-   * Verify a contract.
+   * Verify a contract. In case of delegateCall, check for trustworthiness.
    * @param args - The arguments for the analysis.
    * @param args.chainId - The chain ID.
    * @param args.contract - The contract address.
-   * @returns The analysis result.
+   * @param args.isDelegateCall - Whether the contract is called via delegateCall.
+   * @returns A pair of analysis results: [verification result, delegateCall result (if available)].
    */
   async verifyContract(args: {
     chainId: string;
     contract: Address;
-  }): Promise<ContractAnalysisResult> {
+    isDelegateCall: boolean;
+  }): Promise<[ContractAnalysisResult, ContractAnalysisResult | undefined]> {
```

Comment:
I think we could make it more type-safe like this:
```suggestion
  }): Promise<[ContractAnalysisResult, AnalysisResult<'UNEXPECTED_DELEGATECALL'> | undefined]> {
```

If you also update the signature of `checkDelegateCall` accordingly and use a generic for `mapToAnalysisResult`, it should work:
```ts
  private mapToAnalysisResult<T extends ContractStatus>(args: {
    type: T;
    interactions?: number;
    name?: string;
  }): AnalysisResult<T> {
```

Reply:
[50659a7](https://github.com/safe-global/safe-client-gateway/pull/2720/commits/50659a78acaf8bf4c3c00d2ed1dff66e656b2a40)